### PR TITLE
Inter intra compound

### LINKED ISSUE #329: Compound prediction modes
Trello Reference: https://trello.com/c/w3dmeGQH/45-compound-prediction
Working branch: https://github.com/OpenVisualCloud/SVT-AV1/tree/compound_mode

### PR DESCRIPTION
**Pull request by okhlif**
### Description
Inter intra compound

Closes  #329 

### Author
@chkngit 
@okhlif 

### Type of change
New feature/Tool

### Tests and performance

- BDR gain is 0.24% [28 360p clips face-to-face] 

- Speed cost is about 7% [4 core machine]

### CODE REVIEW

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 1]
```diff
@@ -3560,7 +3850,147 @@ EbErrorType av1_inter_prediction(
...
+            inra_pred_desc.buffer_cr    = intra_pred_cr;
+
+            const int ssx = plane ? 1 : 0;
+            const int ssy = plane ? 1 : 0;
+            const BlockSize plane_bsize = get_plane_block_size(blk_geom->bsize, ssx, ssy);
+            //av1_build_interintra_predictors_sbp
+            uint8_t    topNeighArray[64 * 2 + 1];
+            uint8_t    leftNeighArray[64 * 2 + 1];
+
+#if II_COMP_FLAG
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: there is no need for this FLAG here , it is already new code

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 2]
```diff
@@ -3560,7 +3850,147 @@ EbErrorType av1_inter_prediction(
...
+            uint32_t cu_originy_uv = (pu_origin_y >> 3 << 3) >> 1;
+#else
+            uint32_t cu_originx_uv = (dst_origin_x >> 3 << 3) >> 1;
+            uint32_t cu_originy_uv = (dst_origin_y >> 3 << 3) >> 1;
+#endif
+            if (plane == 0) {
+                dst_ptr = prediction_ptr->buffer_y + prediction_ptr->origin_x + dst_origin_x + (prediction_ptr->origin_y + dst_origin_y) * prediction_ptr->stride_y;
+                dst_stride = prediction_ptr->stride_y;
+                intra_stride = inra_pred_desc.stride_y;
+#if II_COMP_FLAG
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: there is no need for this FLAG here , it is already new code

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 3]
```diff
@@ -3560,7 +3850,147 @@ EbErrorType av1_inter_prediction(
...
+                leftNeighArray + 1,
+                &inra_pred_desc,                                              //uint8_t *dst,
+                                                                                                //int32_t dst_stride,
+                0,                                                                              //int32_t col_off,
+                0,                                                                              //int32_t row_off,
+                plane,                                                                          //int32_t plane,
+                blk_geom->bsize,       //uint32_t puSize,
+                dst_origin_x,
+                dst_origin_y,
+#if II_COMP_FLAG
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: there is no need for this FLAG here , it is already new code

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 4]
```diff
@@ -4937,6 +5367,20 @@ static const int32_t filter_sets[DUAL_FILTER_SET_SIZE][2] = {
 #if COMP_MODE
         candidate_buffer_ptr->candidate_ptr->compound_idx,
         &candidate_buffer_ptr->candidate_ptr->interinter_comp,
+#endif
+#if II_COMP_FLAG
+        &md_context_ptr->sb_ptr->tile_info,
+        md_context_ptr->luma_recon_neighbor_array,
+        md_context_ptr->cb_recon_neighbor_array,
+        md_context_ptr->cr_recon_neighbor_array,
+#if II_COMP_FLAG
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: there is no need for this FLAG here , it is already new code, just leave the comment to not do IFS

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 5]
```diff
@@ -5015,6 +5459,20 @@ static const int32_t filter_sets[DUAL_FILTER_SET_SIZE][2] = {
 #if COMP_MODE
                         candidate_buffer_ptr->candidate_ptr->compound_idx,
                         &candidate_buffer_ptr->candidate_ptr->interinter_comp,
+#endif
+#if II_COMP_FLAG
+                        &md_context_ptr->sb_ptr->tile_info,
+                        md_context_ptr->luma_recon_neighbor_array,
+                        md_context_ptr->cb_recon_neighbor_array,
+                        md_context_ptr->cr_recon_neighbor_array,
+#if II_COMP_FLAG
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: there is no need for this FLAG here , it is already new code

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 6]
```diff
@@ -5092,6 +5550,20 @@ static const int32_t filter_sets[DUAL_FILTER_SET_SIZE][2] = {
...
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: there is no need for this FLAG here , it is already new code

File: Source/Lib/Common/Codec/EbInterPrediction.h [hunk 7]
```diff
@@ -53,6 +53,16 @@ extern "C" {
 #if COMP_MODE
         uint8_t                                compound_idx,
         INTERINTER_COMPOUND_DATA               *interinter_comp,
+#endif
+#if II_COMP_FLAG
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbInterPrediction.h]: better to  enclose all of these information into one struct   e.g  intra_context to reduce the number of input parameters to the MCP call

File: Source/Lib/Common/Codec/EbIntraPrediction.c [hunk 8]
```diff
@@ -4566,6 +4566,84 @@ EbErrorType eb_av1_intra_prediction_cl(
...
+        PLANE_TYPE_Y,                                                                          //int32_t plane,
+        md_context_ptr->blk_geom->bsize,       //uint32_t puSize,
+        md_context_ptr->cu_origin_x,
+        md_context_ptr->cu_origin_y,
+        md_context_ptr->cu_origin_x,                  //uint32_t cuOrgX,
+        md_context_ptr->cu_origin_y,                  //uint32_t cuOrgY
+        0,  //cuOrgX used only for prediction Ptr
+        0   //cuOrgY used only for prediction Ptr
+    );
+
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbIntraPrediction.c]: clean up the commented code

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 9]
```diff
@@ -34,11 +34,56 @@
...
+#endif
 /********************************************
 * Constants
 ********************************************/
+#if II_COMP_FLAG
+// 1 - Regular uni-pred ,
+// 2 - Regular uni-pred + Wedge compound Inter Intra
+// 3 - Regular uni-pred + Wedge compound Inter Intra + Smooth compound Inter Intra
+
+#define II_COUNT                3
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: rename to  e.g  MOTION_MODE_COUNT

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 10]
```diff
@@ -99,6 +144,404 @@ int32_t eb_av1_mv_bit_cost(const MV *mv, const MV *ref, const int32_t *mvjcost,
...
+    PictureControlSet      *picture_control_set_ptr,
+    ModeDecisionContext                  *context_ptr,
+    BlockSize bsize, int bw, int bh,
+    uint8_t* src_buf, uint32_t src_stride, uint8_t* pred_buf, uint32_t pred_stride,
+    int plane_from, int plane_to, int mi_row, int mi_col, int *out_rate_sum,
+    int64_t *out_dist_sum, int *skip_txfm_sb, int64_t *skip_sse_sb,
+    int *plane_rate, int64_t *plane_sse, int64_t *plane_dist);
+static int64_t pick_interintra_wedge(
+    ModeDecisionCandidate        *candidate_ptr,
+    //const AV1_COMP *const cpi,
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: clean up comments

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 11]
```diff
@@ -99,6 +144,404 @@ int32_t eb_av1_mv_bit_cost(const MV *mv, const MV *ref, const int32_t *mvjcost,
...
+    const int bw = block_size_wide[bsize];
+    const int bh = block_size_high[bsize];
+    DECLARE_ALIGNED(32, int16_t, residual1[MAX_SB_SQUARE]);  // src - pred1
+    DECLARE_ALIGNED(32, int16_t, diff10[MAX_SB_SQUARE]);     // pred1 - pred0
+    //if (is_cur_buf_hbd(xd)) {
+    //    aom_highbd_subtract_block(bh, bw, residual1, bw, src->buf, src->stride,
+    //        CONVERT_TO_BYTEPTR(p1), bw, xd->bd);
+    //    aom_highbd_subtract_block(bh, bw, diff10, bw, CONVERT_TO_BYTEPTR(p1), bw,
+    //        CONVERT_TO_BYTEPTR(p0), bw, xd->bd);
+    //}
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: clean up comments

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 12]
```diff
@@ -99,6 +144,404 @@ int32_t eb_av1_mv_bit_cost(const MV *mv, const MV *ref, const int32_t *mvjcost,
...
+    //int64_t skip_sse_sb = INT64_MAX;
+    //int32_t skip_txfm_sb = 0;
+    //int32_t rs = 0;
+    //int64_t rd = INT64_MAX;
+
+
+    //interpolation_filter_search(
+    //                picture_control_set_ptr,
+    //                &pred_desc, //output,
+    //                context_ptr,
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: clean up comments

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 13]
```diff
@@ -99,6 +144,404 @@ int32_t eb_av1_mv_bit_cost(const MV *mv, const MV *ref, const int32_t *mvjcost,
...
+
+    INTERINTRA_MODE best_interintra_mode = INTERINTRA_MODES;
+    EbPictureBufferDesc  inra_pred_desc;
+    inra_pred_desc.origin_x = inra_pred_desc.origin_y = 0;
+    inra_pred_desc.stride_y = bwidth;
+    inra_pred_desc.buffer_y = intrapred;
+
+    int8_t enable_smooth_interintra =1;
+      //if (cpi->oxcf.enable_smooth_interintra &&
+      //!cpi->sf.disable_smooth_interintra) {
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: clean up comments

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 14]
```diff
@@ -99,6 +144,404 @@ int32_t eb_av1_mv_bit_cost(const MV *mv, const MV *ref, const int32_t *mvjcost,
...
+                //if ((!cpi->oxcf.enable_smooth_intra || cpi->sf.disable_smooth_intra) &&
+                //    (INTERINTRA_MODE)j == II_SMOOTH_PRED)
+                //  continue;
+                INTERINTRA_MODE interintra_mode = (INTERINTRA_MODE)j;
+                //rmode = interintra_mode_cost[mbmi->interintra_mode];
+                const int bsize_group = size_group_lookup[context_ptr->blk_geom->bsize];
+                rmode  = candidate_ptr->md_rate_estimation_ptr->inter_intra_mode_fac_bits[bsize_group][interintra_mode];
+#if !II_COMP_FLAG
+                inra_pred_desc.buffer_y = context_ptr->intrapred_buf[j];
+
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: no need for this flag

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 15]
```diff
@@ -619,6 +1062,20 @@ void Unipred3x3CandidatesInjection(
         int16_t to_inject_mv_y = use_close_loop_me ? (inloop_me_context->inloop_me_mv[0][0][close_loop_me_index][1] + BIPRED_3x3_Y_POS[bipredIndex]) << 1 : (me_results->me_mv_array[context_ptr->me_block_offset][list0_ref_index].y_mv + BIPRED_3x3_Y_POS[bipredIndex]) << 1;
         uint8_t to_inject_ref_type = svt_get_ref_frame_type(REF_LIST_0, list0_ref_index);
         if (context_ptr->injected_mv_count_l0 == 0 || mrp_is_already_injected_mv_l0(context_ptr, to_inject_mv_x, to_inject_mv_y, to_inject_ref_type) == EB_FALSE) {
+#if II_COMP_FLAG // 3x3  L0
+             //MvReferenceFrame rf[2];
+             //rf[0] = to_inject_ref_type;
+             //rf[1] = -1;
+
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: clean up comments

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 16]
```diff
@@ -619,6 +1062,20 @@ void Unipred3x3CandidatesInjection(
...
+             uint8_t ii_type;
+             uint8_t is_ii_allowed = 0;//svt_is_interintra_allowed(picture_control_set_ptr->parent_pcs_ptr->enable_inter_intra, bsize, NEWMV, rf);
+             uint8_t tot_ii_types = is_ii_allowed ? II_COUNT : 1;
+             //uint8_t is_obmc_allowed =  obmc_motion_mode_allowed(picture_control_set_ptr, context_ptr->cu_ptr, bsize, rf[0], rf[1], NEWMV) == OBMC_CAUSAL;
+             //tot_ii_types = is_obmc_allowed ? tot_ii_types+1 : tot_ii_types;
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: clean up comments
change ii_type  cand_type

File: Source/Lib/Common/Codec/EbProductCodingLoop.c [hunk 17]
```diff
@@ -1661,6 +1675,13 @@ void set_md_stage_counts(
         context_ptr->md_stage_1_count[CAND_CLASS_3] = (picture_control_set_ptr->slice_type == I_SLICE) ? 0 : (picture_control_set_ptr->parent_pcs_ptr->is_used_as_reference_flag) ? INTER_PRED_NFL : (INTER_PRED_NFL >> 1);
     }
 
+#if II_COMP_FLAG
+#if II_COMP_FLAG
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbProductCodingLoop.c]: no need for flag

File: Source/Lib/Common/Codec/EbProductCodingLoop.c [hunk 18]
```diff
@@ -1680,6 +1701,13 @@ void set_md_stage_counts(
         context_ptr->md_stage_2_count[CAND_CLASS_3] = (picture_control_set_ptr->slice_type == I_SLICE) ? 0 : (picture_control_set_ptr->parent_pcs_ptr->is_used_as_reference_flag) ? 14 : 4;
     }
 
+#if II_COMP_FLAG
+#if II_COMP_FLAG
+    context_ptr->md_stage_2_count[CAND_CLASS_4] = context_ptr->bypass_stage1[CAND_CLASS_4] ? context_ptr->md_stage_1_count[CAND_CLASS_4] : (picture_control_set_ptr->slice_type == I_SLICE) ? 0 : (picture_control_set_ptr->parent_pcs_ptr->is_used_as_reference_flag) ? 12: 4;// 14 : 4;
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbProductCodingLoop.c]: no need for flag

File: Source/Lib/Common/Codec/EbProductCodingLoop.c [hunk 19]
```diff
@@ -1705,6 +1733,13 @@ void set_md_stage_counts(
         context_ptr->md_stage_3_count[CAND_CLASS_3] = (picture_control_set_ptr->slice_type == I_SLICE) ? 0 : (picture_control_set_ptr->parent_pcs_ptr->is_used_as_reference_flag) ? 3 : 1;
     }
 
+#if II_COMP_FLAG
+#if II_COMP_FLAG
+    context_ptr->md_stage_3_count[CAND_CLASS_4] = (picture_control_set_ptr->slice_type == I_SLICE) ? 0 : (picture_control_set_ptr->parent_pcs_ptr->is_used_as_reference_flag) ? 12 : 4;// 14 : 4;
+#else
+    context_ptr->md_stage_3_count[CAND_CLASS_4] = 16;// context_ptr->fast_cand_count[CAND_CLASS_4];//(picture_control_set_ptr->parent_pcs_ptr->is_used_as_reference_flag) ? INTER_PRED_NFL : (INTER_PRED_NFL >> 1);
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbProductCodingLoop.c]: no need for flag

File: Source/Lib/Common/Codec/EbRateDistortionCost.c [hunk 20]
```diff
@@ -1494,7 +1502,39 @@ uint64_t av1_inter_fast_cost(
 
     // NM - To be added when the intrainter mode is adopted
     //  read_interintra_mode(is_compound)
-
+#if II_COMP_FLAG
+    if (md_pass > 0) {
+
+        //MbModeInfo *const mbmi = &cu_ptr->av1xd->mi[0]->mbmi;
+        // inter intra mode rate
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbRateDistortionCost.c]: clean up comments on this block

File: Source/Lib/Common/Codec/EbResourceCoordinationProcess.c [hunk 21]
```diff
@@ -660,6 +660,11 @@ void* resource_coordination_kernel(void *input_ptr)
             sequence_control_set_ptr->enable_altrefs =  sequence_control_set_ptr->static_config.enable_altrefs &&
                 sequence_control_set_ptr->static_config.altref_nframes > 1 &&
                 sequence_control_set_ptr->static_config.encoder_bit_depth == EB_8BIT ? EB_TRUE : EB_FALSE;
+#if II_COMP_FLAG
+            sequence_control_set_ptr->seq_header.enable_interintra_compound = (sequence_control_set_ptr->static_config.enc_mode == ENC_M0) ? 1 : 0;
+
+          //  sequence_control_set_ptr->seq_header.enable_interintra_compound = 1;
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbResourceCoordinationProcess.c]: clean up comments,
shut for 10 bit

File: Source/Lib/Common/Codec/EbModeDecision.c (hunk 9 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbModeDecision.c]: This is still exclusively define the number of Inter-Intra  type to be tested.

**Comment by mafonsoN** [comment]
This doesn't seem like a very good performance vs complexity trade-off. Do you think this tool's compression efficiency or speed could be optimized further at some point?

**Comment by EwoutH** [comment]
At which enc-modes will this feature be enabled with this PR?

**Comment by chkngit** [comment]
this is intended for Mode 0 only for now, the gain on the aom full set is -0.35%.

**Comment by ttrigui** [comment]
@andrey-norkin any comments?

**Pull request merged**